### ISSUE TITLE
Size of serialized object of class recipe

### ISSUE THREAD

**Issue by meztez** [tidymodels/recipes#301]
Hi,

Using the R package recipes to preprocess our data.

Even with the retain=FALSE argument in the prep function. This recipe takes up more than 450 mb when serialized on disk. It is most likely not a bug since I believe it is not optimal to try to serialize a list of objects like a recipe. File size also scales with the size of the data, so I'm still optimistic I can find something.

I told my people as such but they want to be able to use the bake function on the fly in production (single prediction) without having to prep with the original data.

Now, I'm reaching out to you to see if you encountered a similar problem. How would one go about exporting the results of a recipe to later reuse without the size hurtle?

I will keep digging and report back I we find a solution.

Thank you for your time.

```
Data Recipe

Inputs:

      role #variables
        ID          1
  not_used         16
   outcome          1
 predictor         15

Training data contained 2456562 data points and 2340742 incomplete rows. 

Operations:

Variable mutation for  offset [trained]
Log transformation on offset [trained]
Variable mutation for  prod_age [trained]
Orthogonal polynomials on [redacted] [trained]
Variable mutation for  [redacted][trained]
Orthogonal polynomials on [redacted][trained]
Variable mutation for  [redacted][trained]
Interactions with [redacted] [trained]
Mode Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Dummy variable to factor conversion for [redacted][trained]
Dummy variable to factor conversion for [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Factor variables from public_service_ind [trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Mode Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Dummy variable to factor conversion for [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Mean Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Orthogonal polynomials on [redacted][trained]
Mode Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Mean Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Variable mutation for  [redacted][trained]
Factor variables from [redacted][trained]
Mean Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Orthogonal polynomials on [redacted][trained]
Variable mutation for  [redacted][trained]
Orthogonal polynomials on [redacted][trained]
Mean Imputation for [redacted][trained]
Variable mutation for  [redacted][trained]
Variable mutation for  [redacted][trained]
Orthogonal polynomials on [redacted][trained]
Variable mutation for  [redacted][trained]
Variable mutation for  [redacted][trained]
Orthogonal polynomials on [redacted][trained]
Mode Imputation for [redacted][trained]
Factor variables from [redacted][trained]
Dummy variables from [redacted][trained]
Interactions with [redacted] [trained]
Dummy variables from [redacted] [trained]
Interactions with [redacted][trained]
Dummy variables from [redacted][trained]
Interactions with ([redacted]+ [redacted]):[redacted][trained]
Dummy variables from [redacted][trained]
Interactions with ([redacted]+ [redacted]+ [redacted]):[redacted][trained]
Interactions with [redacted]:[redacted][trained]
Check missing values for [redacted], [redacted], [redacted], [redacted], [redacted], [redacted], [redacted], [redacted], [redacted], ... [trained]
Check missing values for [redacted][trained]
```

**Comment by DavisVaughan** [comment]
When you call `recipe()` are you using the full data set as the `data` or `x` argument? This is stored in the recipe in `recipe$template` no matter what. You actually only need a 0-row slice of your data frame here, which has all of the column name and column type information that is required for checking the recipe steps. That can significantly reduce the size of the recipe object.

``` r
suppressPackageStartupMessages({
  library(recipes)
  library(dplyr)
  library(lobstr)
  library(rlang)
})
#> Warning: package 'rlang' was built under R version 3.5.2

# 150,000 rows of iris
big_iris <- bind_rows(rep_along(1:1000, list(as_tibble(iris))))

rec <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris) %>%
  step_center(Sepal.Width)

# take a 0 row slice! (only the types and column names are required at 
# this point)
rec_small <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris[0,]) %>%
  step_center(Sepal.Width)

# quite large!
obj_size(rec)
#> 3,008,104 B

# quite small!
obj_size(rec_small)
#> 8,096 B


# now prep them
p_rec <- prep(rec, big_iris, retain = FALSE)

p_rec_small <- prep(rec_small, big_iris, retain = FALSE)

obj_size(p_rec)
#> 3,014,248 B

obj_size(p_rec_small)
#> 14,240 B
```

<sup>Created on 2019-03-19 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by meztez** [comment]
No we use a 0-row slice. We narrowed down the issue to the dummy step. When serialized, it contains the whole column to be dummified and the training set. I'm still looking into it.

Without the dummy steps, the prepped recipe drops to about 100 ko.

Reporting back

**Comment by meztez** [comment]
In prep.step_dummy, 

I think x$terms should not be used as terms value for step_new_dummy as it returns a list of quosure with enclosed environment.

Only the terms should be returned right? from expr part, as in other step?

I am not sure how this all fits.

[[1]]
<quosure>
expr: ^[redacted]
env:  000000001BA93B38

**Comment by meztez** [comment]
If I use 

`trace("prep.step_dummy", edit=TRUE, where = prep)`

then I edit the line juste before step_dummy_new with

`rm(training, terms)`

I can then save the resulting prepped recipe with saveRDS and it uses significantly less disk space.

Now, I would probably need a little comprehension before being able to do a pull request but I believe the serialization of quosure class object is what is causing me this size issue.

Really interesting

**Comment by meztez** [comment]
This is the resulting prep.step_dummy,

```
function (x, training, info = NULL, ...) 
{
  col_names <- terms_select(x$terms, info = info)
  fac_check <- vapply(training[, col_names], is.factor, logical(1))
  if (any(!fac_check)) 
    warning("The following variables are not factor vectors and will be ignored: ", 
      paste0("`", names(fac_check)[!fac_check], "`", collapse = ", "), 
      call. = FALSE)
  col_names <- col_names[fac_check]
  if (length(col_names) == 0) {
    stop("The `terms` argument in `step_dummy` did not select ", 
      "any factor columns.", call. = FALSE)
  }
  levels <- vector(mode = "list", length = length(col_names))
  names(levels) <- col_names
  for (i in seq_along(col_names)) {
    form_chr <- paste0("~", col_names[i])
    if (x$one_hot) {
      form_chr <- paste0(form_chr, "-1")
    }
    form <- as.formula(form_chr)
    terms <- model.frame(form, data = training, xlev = x$levels[[i]], 
      na.action = na.pass)
    levels[[i]] <- attr(terms, "terms")
    attr(levels[[i]], "values") <- levels(getElement(training, 
      col_names[i]))
    **attr(levels[[i]], ".Environment") <- NULL**
  }
  **rm(col_names, fac_check, training)**
  step_dummy_new(terms = x$terms, role = x$role, trained = TRUE, 
    one_hot = x$one_hot, naming = x$naming, levels = levels, 
    skip = x$skip, id = x$id)
}
```

We will conduct other test to see if this breaks some stuff.

**Comment by DavisVaughan** [comment]
oh! its the terms object. As you saw, it has an embedded environment.

``` r
suppressPackageStartupMessages({
  library(recipes)
  library(dplyr)
  library(lobstr)
  library(rlang)
})
#> Warning: package 'rlang' was built under R version 3.5.2

# 150,000 rows of iris
big_iris <- bind_rows(rep_along(1:1000, list(as_tibble(iris))))

rec <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris[0,]) %>%
  step_dummy(Species)

p_rec <- prep(rec, big_iris, retain = FALSE)

# it says they are equal, because much of the 
# memory is duplicated and shared between them
obj_size(p_rec)
#> 6,540,360 B

obj_size(p_rec$steps[[1]]$levels$Species)
#> 6,540,360 B

embedded_env <- attr(p_rec$steps[[1]]$levels$Species, ".Environment")

env_names(embedded_env)
#>  [1] "terms"           "form"            "form_chr"       
#>  [4] "i"               "levels"          "fac_check"      
#>  [7] "col_names"       ".Generic"        ".Class"         
#> [10] ".Method"         ".Group"          ".GenericCallEnv"
#> [13] ".GenericDefEnv"  "x"               "training"       
#> [16] "info"            "..."

attr(p_rec$steps[[1]]$levels$Species, ".Environment") <- NULL

obj_size(p_rec)
#> 1,124,696 B

# the resulting size comes from storing
# then recipes package env in the function
# here, which could also be fixed i think
obj_size(p_rec$steps[[1]]$naming)
#> 1,109,232 B
```

<sup>Created on 2019-03-19 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by meztez** [comment]
yeah I don't think this is good, I'm suggesting a fix

**Comment by topepo** [comment]
This is a specific thing that an intern is working on this summer. We're going to create cleaning functions to get rid of these extra bits that you _may_ not need after the prepping.

**Comment by DavisVaughan** [comment]
Yea, the only things in _this_ particular formula that creates the terms object would be a single column name that should _always_ be looked up in the data itself. So we have no need to the env. It would be safe to nuke this one. 

As @topepo says, a more general approach will also be a summer intern project

**Comment by DavisVaughan** [comment]
I think that the only required patch for this right now would be to add

```r
attr(levels[[i]], ".Environment") <- NULL
```

directly after this line

https://github.com/tidymodels/recipes/blob/53b3b298bfcee27a1098494e62613b7f099e376b/R/dummy.R#L184

as @meztez did in the example

**Comment by topepo** [comment]
Keep in mind that the quosures in the `term` objects might also capture the global env

**Comment by meztez** [comment]
we tested internally and it seems to be fine for the quosures, it might be related to the way R serialization is done. Pull request done.

**Comment by meztez** [comment]
Resolved with #304
